### DISCUSSION TITLE
Road to spec decorator

### DISCUSSION THREAD

**Discussion by Austaras** [swc-project/swc#5053] [Ideas]
### Some history
There are historically three versions of decorator
1. legacy decorator. Babel and tsc both support it, and it gains widely adopted in js community.
2. static decorator. Some engine people believe version 1 is too dynamic to optimize, so they came up with this. Nobody actually uses it, and babel would drop it in version 8.
3. current decorator. It's already entered stage 3. Babel supports it as `2021-12` and would be default in version 8. TS and esbuild people have already been working on it. It shares many similarities with 1.

### Current status of swc
There're currently two options related with decorator in swc, which is `jsc.parsre.decorators` and `jsc.transform.legacyDecoratos`. Enable former would enable decorator in parser, and static decorator in transform. Enable later would enable legacy decorator.

This poses a problem in how to expose API for spec decorator.

### Possible solution
1. release a new major version to drop support for static decorator, and substitute it with spec decorator. This sounds a bit overreacting, but considering the many problems of current swc js config, it is not entirely unacceptable.
2. add an option in transform, maybe called `specDecorators`, and if both this and `legacyDecoratos` is set to true, throw an error. It has no other downside than ugly.

**Comment by kdy1** [comment]
I think the second option is better, as I want to do much more cleanup (of config) for the next major version release

**Reply by Austaras** [reply]
Oh, glad to know there'a alreay a plan to clean config

**Reply by Austaras** [reply]
First ts version that support spec decorator would be 4.8 and it will be release in august. If rework of config could be done before that, solution 1 is still possible

**Reply by kdy1** [reply]
Problem is that we want to do much more than config cleanup for v2.
One of planned item is stabilization of plugins (with some compat)

**Comment by Austaras** [comment]
Well tsc people hasn't started working on decorator yet...

**Comment by yay** [comment]
Looks like they have finished:
https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#decorators
https://github.com/evanw/esbuild/issues/104#issuecomment-1414537184

**Comment by Amareis** [comment]
So, TypeScript take different approach - new decorators is available until you specify `legacyDecorators: true`. Maybe swc should use such strategy too? It's much more straightforward, aligned with ES spec, which obviously will became standart, and doesn't need any additional flags in config.

**Reply by Austaras** [reply]
TSC can freely do break change while SWC cannot -- SWC would generate old spec decorator if `legacyDecorators` is not specified, although I haven't seen any people using it.

**Reply by kdy1** [reply]
Honestly, I don't expect someone to use it, but I'll implement it with an option at start

**Comment by kdy1** [comment]
I started working at https://github.com/swc-project/swc/pull/6950

**Comment by graham-fisher-ii** [comment]
Hello! Saw that [#6950](https://github.com/swc-project/swc/pull/6950) was included in 1.3.45: thanks for your hard work! However, I am having trouble using the stage 3 spec decorators feature.

-------

When I copy the [simple example from the TS 5 beta announcement](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#decorators) and configure `.jsc.parser.syntax: "typescript"`, `.jsc.parser.decorators: true`, and `.jsc.transform.legacyDecorator: false`, the output emitted looks like it's still expecting the legacy decorator interface:
```
var __decorate = (void 0) && (void 0).__decorate || function(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
```
See [this playground](https://play.swc.rs/?version=1.3.46&code=H4sIAAAAAAAAA32Rz0rEMBDG732KoXhoYckDtAheRA%2BK4lXEDdnZbCBNSjLFLcu%2Bu%2Fm3ZStqLmEyM99vvsl%2BMoKUNaCtlLh7RjrYXWOdkspwncMOuJk38CmsITxSCls4VRWEs78IOBw1FzigoaJCB%2BVLL2OMO5mj94%2FYDOUEUW81ssBv6qeXhw7uA8UpI2FIMqxuV8UUSH7SBLewHpMJrnWCLry2%2F49zVPQ7xiFNzhROljhnt0vmh9e%2BCgVCc%2B%2FhFZ0P68gODR%2BwA0%2FRTxZKFtwkyLrmOnu9k%2BiBxWTwGK%2FVCHfXP5VepEOk5q%2Blbh9Ra7uBYU5aoDzcnBbCmW3bi3y0kBY8Ri5%2BFStN%2FcbnOlSNrJD66htLW%2BEBNwIAAA%3D%3D&config=H4sIAAAAAAAAA1VQzQ6DIAy%2B%2BxSmZ6%2B77LzrHoJgNRigpMVkxPjuAydObm2%2F33Tr%2Bh4W0fDstzzmJSgW5GvPF0k%2Bqk%2B%2BQEwBRbMJEYaKjqiJVSSWzIi84oVEKaJJWcHjtP8QiKy8TMTuHmJxVjq9qlkVDq1S8YyxNEF5nBXAEgm2fHDGmynd%2FTW5wCjSEgtV%2BdliW7Q7I8HRuB7g%2BZzygBKf3Rz5ReDPrImXOxh5V3n5S7d%2FAbUn%2BChtAQAA). Running the output code in Google Chrome (Version 111.0.5563.146 (Official Build) (arm64)) does not yield the expected logging: 
![image](https://user-images.githubusercontent.com/49082176/230255654-562b9b53-6519-4110-bec5-cf8d904967e7.png)

-------

I also tried try removing the types and configuring `.jsc.parser.syntax: "ecmascript"`, `.jsc.parser.decorators: true`, and `.jsc.transform.legacyDecorator: false`. The output emitted looks like it calls the decorators with a single argument, an `elementObject` derived from the `elementDescriptor` derived from an element's `def`: the `elementObject` looks somewhat similar to the [decorator proposal's context](https://github.com/tc39/proposal-decorators#class-methods), but is not quite the same:
```
var elementFinisherExtras = _toElementFinisherExtras((0, decorators[i])(elementObject) || elementObject);
```
See [this playground](https://play.swc.rs/?version=1.3.46&code=H4sIAAAAAAAAA32RTWrDMBCF9z7FYLKQwegAEYVuSrtoaekFGqFMFMFYMtKYJoTcvfJfGoem2mn05n1vRrvOG3bBAwVrcfuGvA9bEaKzzmsarzV8meAZD1zBqYB88jUFQpmb7mvVIN3NgIgtaYMNep4oUkodbZpNb43L1%2FfnNTxls%2Bi8hWZokmW1EHP2TR0xPMAyiDSaSPDepRpmkPqPc3D8NyYid9FPnNHiXBSLl5vJVJEFhnRK8IEx5eHHCb1uUF0WyLEzHKLoq9c76DPLvphn%2Bu2YkI%2FX%2FzRUbERkcW%2BJmxckCjU0x8ELXILV6UI4y0012%2FeRh4W2PRe%2Fp%2Bii%2FNTHMqtaOZFU8QPyJIIKNQIAAA%3D%3D&config=H4sIAAAAAAAAA1VQOw6DMAzdOQXyzNqlc9cewgoGBeUnO0iNEHdvQgklm9%2FXlreu72ERBc9%2By2MGAVmIL5wZSS7iJzNAyqIo1iHCUNWRlGeMniU7Iq90KYuU0IRG6KD2nwKR0cnk2d6XGJpRpVctq8GhTSLPFI9L5HGeAMZ7odYPVjs9pXu%2F8jYwibTGYkU3G2oP7c6VYP24HuL5nJhCQaXNercI%2FJ1149UOWt41Xv7S7V8m3JykbQEAAA%3D%3D): running the output code in Google Chrome (Version 111.0.5563.146 (Official Build) (arm64)) yields an error:
![image](https://user-images.githubusercontent.com/49082176/230255819-2295558a-0d43-4efd-9fbd-16927d7cc209.png)

-------

Are these examples expected to work? Am I missing some configuration needed to use spec decorators with swc? Am I (and the TS 5 beta announcement) mistaken about the spec decorator API or another detail of how this is supposed to work? Is this the right forum to discuss this?

**Reply by miso-belica** [reply]
I though stage 3 decorators are not supported by SWC but @kdy1 in https://github.com/swc-project/swc/issues/3062#issuecomment-1498999769 says they are. Now I checked it and it seems legacy decorator structure is still emitted by swc. **Is there anything what I have to do to emit stage 3 proposal for decorators** the same way as TS v5 emits? I am using @swc/jest and I have the test below. As @tgfisher4 mentions I see in the debugger old structure is emitted.

![image](https://user-images.githubusercontent.com/1072330/230391227-8714c9dd-2de1-4275-bdfa-7659f49890c6.png)

```json5
// .swcrc
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "decorators": true,
    },
    "transform": {
      "legacyDecorator": false
    }
  }
}
```

```ts
const _BOUND_FUNCTIONS = {};

export class BackgroundJob {
  static readonly bind = <This, Args extends unknown[], Return extends void>(
    method: (this: This, ...args: Args) => Promise<Return>,
    context: ClassMethodDecoratorContext<This, (this: This, ...args: Args) => Promise<Return>>,
    ...argss: any[]
  ) => {
    const name = String(context.name);
    context.addInitializer(function () {
      _BOUND_FUNCTIONS[name].bind(this, argss);

      this[context.name] = async (...args: Args): Promise<void> => {
        await BackgroundJob.create({name, args});
      };
    });
  };
}

class TestClass {
  calls: string[] = [];

  @BackgroundJob.bind
  success(s: string, n: number) {
    this.calls.push(`success(${s}, ${n})`);
    return Promise.resolve();
  }
}
```

**Reply by kdy1** [reply]
Currently, there's no JS api. It's only implemented
